presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java

```diff
@@ -258,18 +262,37 @@ private int readNoFilter(int[] positions, int positionCount)
             throws IOException
     {
         // filter == null implies outputRequired == true
-        if (positions[positionCount - 1] == positionCount - 1) {
+        int maxPositionCount = positions[positionCount - 1] + 1;
+
+        if (maxPositionCount * (1 - BATCH_EXECUTION_FILTERRATE_THRESHOLD) <= positionCount) {
             if (presentStream != null) {
-                int nonNullCount = positionCount - presentStream.getUnsetBits(positionCount, nulls);
+                int nonNullCount = maxPositionCount - presentStream.getUnsetBits(maxPositionCount, nulls);
                 dataStream.next(values, nonNullCount);
-                unpackByteNulls(values, nulls, positionCount, nonNullCount);
+
+                if (maxPositionCount != positionCount) {
```

Comment:
nit: perhaps, use > instead of != to clarify that maxPositionCount can only be >= positionCount, but not <; ditto all other places